[stdlib] Set, Dictionary: Take the max load factor into account in `.init(minimumCapacity:)`

lorentey opened the pull request:
This fixes `init(minimumCapacity:)` on `Set` and `Dictionary` to take the maximum load factor into account before allocating storage. Previously, these initializers usually resulted in collections with less than the specified minimum capacity.  (`reserveCapacity(_:)` does not have the same issue.)

rdar://problem/36619317

lorentey committed:
[stdlib] Set, Dictionary: Take the max load factor into account in .init(minimumCapacity:)

rdar://problem/36619317

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please benchmark

natecook1000 reviewed (approved):
Great catch on this! Only feedback I have is to ask whether it would make sense to rename `_NativeBuffer(minimumCapacity:)`, since it enforces a different guarantee than this init. Otherwise, LGTM!

lorentey committed:
[stdlib] Set, Dictionary terminology: Distinguish capacity from bucket count.

"Capacity" in the public Set/Dictionary API means the maximum number of items that a collection instance may hold without reallocation. In implementation details, the same word was also used for the capacity of the underlying storage buffer, which is usually higher due to the maximum load factor.

Having the same name for two different values has led to bugs where the logical/effective capacity is passed directly to a function expecting the raw storage capacity, or vice versa. This change updates internal names to clearly distinguish between the two senses, introducing "bucketCount" as the name for the storage buffer capacity, and reserving "capacity" for the maximum logical item count.

lorentey committed:
[test] Use the public capacity in tests, not the internal bucket count.

natecook1000 commented:
@lorentey 💯💯💯

lorentey commented:
@swift-ci please test

lorentey commented:
@natecook1000 That was a great suggestion; switching to better names uncovered another case of the same bug in lurking in `Dictionary.mapValues(_:)`.
